### ISSUE TITLE
Integration of igraph_get_shortest_path_astar

### ISSUE THREAD

**Issue by JO-WTF** [igraph/python-igraph#652]
Hi there,

I'm wondering if there are any plans to add `igraph_get_shortest_path_astar` to python-igraph.

I'm working on an algorithm for AGV path-planning using the A-star algorithm. To reduce turns, I plan to modify the heuristics function. It seems like the `igraph_get_shortest_path_astar` is a perfect match for my needs.

I came across issue #576 , which discusses integrating `get_k_shortest_paths`, but it seems like it might be too challenging for me as I've never tried to develop interfaces between C and python. Additionally, `get_shortest_path_astar` requires a heuristic parameter, which could make things even more complicated.

So I'd like to request this feature. At the same time, I'm happy to try it out if you could provide more detailed instructions.

Thanks a lot for your help.

**Comment by ntamas** [comment]
> I'm wondering if there are any plans to add igraph_get_shortest_path_astar to python-igraph.

Yes, there are, thanks for the feature request. I'll try to find some time for it tonight or tomorrow. Will you be able to compile a new version of `python-igraph` from source using the Github repo?

**Comment by JO-WTF** [comment]
> 

Thanks a lot for the quick support. Yes I am sure I can manage to do that.

**Comment by ntamas** [comment]
Okay, commit 22484a9 adds support for this function in `Graph.get_shortest_path_astar()`. Can you please test it a bit to see if it satisfies your needs?

A short usage example:

```python
n = 4
g = Graph.Lattice((n, n), circular=False)
xs, ys = list(range(n)) * n, sum(([i] * n for i in range(n)), [])
g.vs["coord"] = list(zip(xs, ys))

def heuristics(graph, u, v):
    ux, uy = graph.vs[u]["coord"]
    vx, vy = graph.vs[v]["coord"]
    return ((ux - vx) ** 2 + (uy - vy) ** 2) ** 0.5

print(g.get_shortest_path_astar(0, 11, heuristics))
```

**Comment by szhorvat** [comment]
@ntamas I only have my phone so can't check how this is working but I was wondering if it'd make sense to allow passing in Euclidean point sets, or even a pre-computed distance matrix, for better performance, and for avoiding too many Python function calls.

**Comment by JO-WTF** [comment]
@ntamas , I have tested it and it works brilliantly. Using Manhattan distance as a heuristic, it runs 3.25 times faster than `get_shortest_paths`.

However, it is worth mentioning that in order to check the current direction, the current path needs to be a parameter of heuristics as well. This allows for penalties to be added when the next node is not in the correct direction.

Perhaps we could consider making current path an optional parameter of heuristics? This would allow users to customize their own heuristics with more flexibility.

But I guess that would require more work on the core igraph code instead of only an interface?

**Comment by szhorvat** [comment]
This is an experimental function in the C core so changes to the interface are possible.

Can you please open a new issue in the C core repo and explain the need for the path parameter in detail. Include references to articles discussing using the existing path, or other software sounds the same, if you have any.

**Comment by JO-WTF** [comment]
> 

Sure, I am doing it now.

**Comment by JO-WTF** [comment]
Hi, @szhorvat ， I have submitted a feature request here https://github.com/igraph/igraph/issues/2327

**Comment by JO-WTF** [comment]
@szhorvat Ah，I'm really sorry, I just realized that I made a mistake. Actually, what I need to modify is the method for evaluating `g_cost`. Assuming that I just traveled from the point (-1,0) to (0,0), and the adjacent nodes are (1,0), (0,1), and (0,-1), when evaluating the `g_cost` of these three neighbors, I need to check whether they are consistent with the current traveling direction (1,0). The `h_cost` will be just the Manhattan distance and needs no path info.

**Comment by vlund** [comment]
> @ntamas I only have my phone so can't check how this is working but I was wondering if it'd make sense to allow passing in Euclidean point sets, or even a pre-computed distance matrix, for better performance, and for avoiding too many Python function calls.

I would love the possebility to pass N-dimensional point sets (a pre-computed distance matrix would probably be too expensive in my case due to the quadratic space complexity).

**Comment by ntamas** [comment]
Actually, I think this is already possible with a custom class instance; something like:

```python
class DistanceCalculator:
    def __init__(self, points):
        self.points = points
        
    def __call__(self, graph, u, v):
        return ...    # calculate the distance here
        
heuristics = DistanceCalculator(points)
path = g.get_shortest_path_astar(source, target, heuristics)
```

**Comment by szhorvat** [comment]
@ntamas Would you support implementing optimized versions of distances between numeric vectors, i.e. "coordinates", both for different metrics and different dimensions?

@vlund If @ntamas likes the idea, do you have the time an expertise to contribute something like this to the C core?

**Comment by vlund** [comment]
> This is not correct as the heuristic should underestimate, not overestimate the actual distance. You'd need Euclidean distance, not Manhattan.

Ah, thanks! It should indeed be a straight-line heuristic, like Euclidean distance, as you mentioned.

@szhorvat Sorry, I can't say that I have much experience with coding in C.  :/

**Comment by ntamas** [comment]
See [my comment](https://github.com/igraph/igraph/issues/2327#issuecomment-1676320715) over there in the igraph/igraph repo. TL;DR: I have come to the conclusion that the A* algorithm in the C core is not really useful as is because the vast majority of problems that one might want to solve with A* cannot be modelled on a static graph that is generated in advance. To make things usable one would need to implement A* on a more abstract level with state spaces, actions and dynamically generated neighbor states, as I had to do it myself in some other project where I needed to work with A* recently.

To this end, I'm closing the current issue. If there will ever be a better (more abstract) A* implementation in C core, it will be much easier to implement custom heuristics like Euclidean distance on top of it. I do not plan to move forward with the current implementation in the C core without severe modifications.
